nhs/data/join_on_sa1.py

```diff
@@ -0,0 +1,219 @@
+from typing import Any, List  # Importing required types
+
+import fiona
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import numpy as np
+import polars as pl
+import yaml
+from fiona.drvsupport import supported_drivers
+from loguru import logger
+from pathos.multiprocessing import ProcessingPool as Pool
+from pyproj import CRS
+
+# Set the SHAPE_RESTORE_SHX option
+supported_drivers["ESRI Shapefile"] = "rw"
+
+
+def read_config(config_path: str) -> dict:
+    """Read YAML configuration file."""
+    with open(config_path, "r") as file:
+        config = yaml.safe_load(file)
+    logger.info(f"Configuration loaded: {config}")
+    return config
+
+
+def read_shapefile(path: str) -> int:
+    """Read shapefile and return total number of rows."""
+    total_rows = len(pl.read_parquet(path))
+    logger.info(f"Total number of rows in the shapefile: {total_rows}")
+    return total_rows
+
+
+def read_csv(path: str) -> pl.DataFrame:
+    """Read CSV file and return as Polars DataFrame."""
+    df = pl.read_csv(path)
+    logger.info(df.head())
+    return df
+
+
+def prepare_points(df: pl.DataFrame, crs: str) -> Any:
+    """Prepare points data for geospatial analysis."""
+    if "LONGITUDE" in df.columns and "LATITUDE" in df.columns:
+        df = df.rename({"LONGITUDE": "x", "LATITUDE": "y"})
+
+    gdf = gpd.GeoDataFrame(
+        df.to_pandas(),
+        geometry=gpd.points_from_xy(df["x"], df["y"]),
+        crs="EPSG:7844",  # Assuming input is in WGS84
+    )  # type: ignore
+    projected_crs = CRS.from_string(crs)
+    return gdf.to_crs(projected_crs)
+
+
+def process_chunk(map_data: Any, chunk: Any) -> pl.DataFrame:
+    """Process a chunk of point data, joining with area data."""
+    pnts_with_area = gpd.sjoin(chunk, map_data, how="left", predicate="within")
+    missing_points = pnts_with_area[pnts_with_area["SA2_NAME21"].isnull()]
+
+    if not missing_points.empty:
+        nearest_join = gpd.sjoin_nearest(
+            missing_points[["geometry"]], map_data, how="left"
+        )
+        pnts_with_area.loc[missing_points.index, "SA2_NAME21"] = nearest_join[
+            "SA2_NAME21"
+        ]
+        pnts_with_area.loc[missing_points.index, "SA1_CODE21"] = nearest_join[
+            "SA1_CODE21"
+        ]
+
+    result = pnts_with_area[["x", "y", "SA2_NAME21", "SA1_CODE21"]]
+    result = result.rename(columns={"SA2_NAME21": "area", "SA1_CODE21": "area_code"})
+    return pl.DataFrame(
+        result.to_dict()
+    )  # Convert to dict first to ensure compatibility with Polars
+
+
+def parallel_process(pnts_gdf: Any, map_data: Any, num_cores: int) -> pl.DataFrame:
```

Comment:
Types would be geopandas dataframe and polars dataframe?

Reply:
Also is writing a custom parallel function needed? If you can manage to keep everything as `LazyFrame`s without ever calling `collect()` or converting to `DataFrame` then `polars` will automatically split the processing across several CPU cores. Is it not possible to convert geopandas frames to polars frames?

Reply:
Types corrected.

There is a geopolars lib but I don't know how compatible it is with the other geo-related libraries I've used. There is a way to convert to DFs but it's pretty complicated and requires transforming points. 

There is the geopolars library but I don't know how interchangeable it is with geopandas, I think for now I'll have it in geopandas and if it's too slow I'll go to geopolars.